clang/lib/Format/NumericLiteralCaseFixer.cpp

```diff
@@ -0,0 +1,347 @@
+      }
+      return C == Transforms.Separator || llvm::isDigit(C);
+    });
+  }
+  // The next character of a floating point literal will either be the
+  // separator, or the start of a suffix.
+  FloatExponentSeparatorBegin = Cur;
+  if (IsFloat) {
+    const char LSep = IsHex ? 'p' : 'e';
+    const char USep = IsHex ? 'P' : 'E';
+    Cur = std::find_if_not(
+        Cur, End, [LSep, USep](char C) { return C == LSep || C == USep; });
+  }
+  FloatExponentSeparatorEnd = Cur;
+
+  // Fast forward through the exponent part of a floating point literal.
+  if (!IsFloat) {
+  } else if (FloatExponentSeparatorBegin == FloatExponentSeparatorEnd) {
+  } else {
+    Cur = std::find_if_not(Cur, End, [](char C) {
+      return llvm::isDigit(C) || C == '+' || C == '-';
+    });
+  }
+
+  // Find the range containing a suffix if any.
+  SuffixBegin = Cur;
+  size_t const SuffixLen = End - Cur;
+  StringRef suffix(&(*SuffixBegin), SuffixLen);
+  if (!matchesReservedSuffix(suffix)) {
+    Cur = std::find_if_not(Cur, End, [](char C) {
+      // In C++, it is idiomatic, but NOT standard to define user-defined
+      // literals with a leading '_'. Omit user defined literals from
+      // transformation.
+      return C != '_';
+    });
+  }
+  SuffixEnd = Cur;
+}
+
+void QuickNumericalConstantParser::applyFormatting() {
+
+  auto Start = Formatted.cbegin();
+  auto End = Formatted.cend();
+
+  assert(Start <= PrefixBegin && End >= PrefixBegin &&
+         "PrefixBegin is out of bounds");
+  assert(Start <= PrefixEnd && End >= PrefixEnd &&
+         "PrefixEnd is out of bounds");
+  assert(Start <= HexDigitBegin && End >= HexDigitBegin &&
+         "HexDigitBegin is out of bounds");
+  assert(Start <= HexDigitEnd && End >= HexDigitEnd &&
+         "HexDigitEnd is out of bounds");
+  assert(Start <= FloatExponentSeparatorBegin &&
+         End >= FloatExponentSeparatorBegin &&
+         "FloatExponentSeparatorBegin is out of bounds");
+  assert(Start <= FloatExponentSeparatorEnd &&
+         End >= FloatExponentSeparatorEnd &&
+         "FloatExponentSeparatorEnd is out of bounds");
+  assert(Start <= SuffixBegin && End >= SuffixBegin &&
+         "SuffixBegin is out of bounds");
+  assert(Start <= SuffixEnd && End >= SuffixEnd &&
+         "SuffixEnd is out of bounds");
+
+  std::transform(PrefixBegin, PrefixEnd, PrefixBegin, Transforms.Prefix);
+  std::transform(HexDigitBegin, HexDigitEnd, HexDigitBegin,
+                 Transforms.HexDigit);
+  std::transform(FloatExponentSeparatorBegin, FloatExponentSeparatorEnd,
+                 FloatExponentSeparatorBegin,
+                 Transforms.FloatExponentSeparator);
+  std::transform(SuffixBegin, SuffixEnd, SuffixBegin, Transforms.Suffix);
+}
+
+std::optional<std::string> QuickNumericalConstantParser::formatIfNeeded() && {
+  parse();
+  applyFormatting();
+
+  return (Formatted == IntegerLiteral)
+             ? std::nullopt
+             : std::make_optional<std::string>(std::move(Formatted));
```

Comment:
```suggestion
             : std::move(Formatted);
```
Doesn't this suffice?